### ISSUE TITLE
[Ionic React] Sampler does not work in mobile environments (iOS/Android)

### ISSUE THREAD

**Issue by ShayanJavadi** [Tonejs/Tone.js#556]
I'm using [Ionic React](https://ionicframework.com/blog/announcing-ionic-react/) to build my React app for iOS and Android. I was pretty excited about this new integration of React in Ionic because I can utilize Web Audio in mobile without having to rewrite my app.

Only issue right now is that my bread and butter Sampler does not work. I'm using Sampler to create a piano from a couple of samples. 

It works just fine on the web, but it looks like Tone has issues loading the sample files into the buffer on mobile. The files are definitely there, and I've gone in the Tone.js code and have verified this. 

The root of the issue from me digging in the source code looks to be after the buffer for the file is 95% loaded. I'm not able to figure out how the buffer management stuff works but I hope that this is a start. I also noticed that the buffer has a `readyState` of  4, and a `status` of 0.

Any help on this would be appreciated. I can't provide a demo since you'd have to build this to iOS and then run it with Xcode. But I can provide the source code for my app that I'm working on:

Here's the repo: https://github.com/ShayanJavadi/solfej

[This is the line that the sampler gets used ](https://github.com/ShayanJavadi/solfej/blob/0cae29154529a5ab5e31d80e2c6974e63ae0c7e2/src/store/instrument/actions.js#L45)

Screenshot of the error:
<img width="977" alt="Screen Shot 2019-10-18 at 12 41 38 PM" src="https://user-images.githubusercontent.com/11575429/67116191-900a4200-f1a5-11e9-8499-45cd38d13071.png">
Screenshot of the iOS build folder where we can verify that the samples are present.
<img width="363" alt="Screen Shot 2019-10-18 at 12 41 57 PM" src="https://user-images.githubusercontent.com/11575429/67116193-900a4200-f1a5-11e9-81ac-fb44522dd046.png">

And again any help would be appreciated. Been busting my butt on this issue for the past couple of days as it is crucial to the success of my app.

**Comment by tambien** [comment]
Have you taken a look at [this wiki page](https://github.com/Tonejs/Tone.js/wiki/Using-Tone.js-with-React-or-Vue)? Similar to the suggestions in the wiki, [this SO answer](https://stackoverflow.com/questions/44154939/load-local-images-in-react-js) suggests either using `import` to handle the url importing or use the public folder. Let me know if either of those works for you

**Comment by ShayanJavadi** [comment]
@tambien Yup that's [exactly how I'm handling imports right now.](https://github.com/ShayanJavadi/solfej/blob/0cae29154529a5ab5e31d80e2c6974e63ae0c7e2/src/store/instrument/actions.js#L45). 

Like I said the importing doesn't seem to be the issue. From what I understand from the code the buffer would not go to 95% completion.

Any ideas?

**Comment by Seanitzel** [comment]
I'm using Sampler for my project, which is a PWA made using Vue and it works great on all environments, so you might wanna check other things that might cause this.

Also, something you can try is loading the buffers yourself and then sending them to the Sampler using the add() method instead of handing out url's to the sampler.

**Comment by ShayanJavadi** [comment]
@Seanitzel My react app is working fine on all platforms as well. It’s the iOS/android build that’s giving me the issue. Not web iOS, native.

Do you happen to have an example anywhere on how to use the add() method? Is that part of the Sampler?

Appreciate your help

**Comment by ShayanJavadi** [comment]
I was able to solve this by manually fetching the sample files:

```
  const A4 = await fetch("capacitor://localhost/A4.mp3")
  const A4Blob = await A4.blob()

  const A5 = await fetch("capacitor://localhost/A5.mp3")
  const A5Blob = await A5.blob()

  const A6 = await fetch("capacitor://localhost/A6.mp3")
  const A6Blob = await A6.blob()
  
  // console.log(a, b, URL.createObjectURL(b))

  piano = new Sampler(
    {
      A4: URL.createObjectURL(A4Blob),
      A5: URL.createObjectURL(A5Blob),
      A6: URL.createObjectURL(A6Blob)
    },
    {
      onload: () => onLoadFunction
    }
  ).toMaster();
```

Looks like Tone's Buffer does not work well with iOS native build files for some reason.

**Comment by Seanitzel** [comment]
Great, 

In case you wanna add more sounds to the sampler after instansiating it - the add method is indeed part of the Sampler api: https://tonejs.github.io/docs/13.8.25/Sampler#add

An example:
```
const file = await fetch(url)
const buffer = file.arrayBuffer()
const audioBuffer = Tone.context.decodeAudioData(buffer)
sampler.add('c3', audioBuffer)
```

**Comment by ranbena** [comment]
I ran into the same problem in which iOS raised an error when providing the sampler with relative paths to sound files.
Solved by utilizing  @Seanitzel's suggestion like so:

```javascript
import { Sampler } from "tone";

const data = {
  A1: require('../assets/sounds/A1.mp3'),
  B1: require('../assets/sounds/B1.mp3'),
  ...
};

class Example {
  sampler: Sampler;

  constructor() {
    this.sampler = new Sampler().toDestination();

    const promises = Object.entries(data).map(this.addToSampler);
    Promise.all(promises).then(() => {
      console.log('done');
    });
  }

  addToSampler = ([note, path]) =>
    fetch(path)
      .then(file => file.arrayBuffer())
      .then(buffer => context.decodeAudioData(buffer))
      .then(audioBuffer => {
        this.sampler.add(note, audioBuffer);
      });
}
```

Thank you @ShayanJavadi  and @Seanitzel 🙇
